showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ImpressionCollector.kt

```diff
@@ -0,0 +1,55 @@
+package com.vinaygaba.showkase_component_impressions
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+
+internal class ImpressionCollector<T>(
+    val coroutineScope: CoroutineScope,
+    val onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    private val mutableImpressionEvents = MutableStateFlow<ShowkaseVisibilityEvent<T>?>(null)
+
+    @OptIn(FlowPreview::class)
+    val impressionEvents: Flow<ShowkaseVisibilityEvent<T>> =
+        mutableImpressionEvents.filterNotNull().debounce(200)
+
+    internal fun onLayoutCoordinatesChanged(
+        key: T,
+        visibilityPercentage: VisibilityMetadata,
+    ) {
+        publishImpressionEvent(key, visibilityPercentage)
+    }
+
+    internal fun publishImpressionEvent(
```

Comment:
it's confusing that some of these "publish*" functions actually call back to `onVisibilityEvent`, and some emit to the flow, which triggers another call down the line. it is hard to follow

Reply:
I hear you and this was important to do to get around an issue that I spent a couple days debugging. Since I am using the coroutineScope that's tied to the composition (using `rememberCoroutineScope`), it causes the coroutine to be called. Thus, I cannot invoke a coroutine from the onDispose method and was not able to publish to the flow directly.